scripts/gke/util.sh

```diff
@@ -233,27 +233,17 @@ gcpGenerateKsApp() {
   ks generate cloud-endpoints cloud-endpoints
   ks generate cert-manager cert-manager --acmeEmail=${EMAIL}
   ks generate iap-ingress iap-ingress --ipName=${KUBEFLOW_IP_NAME} --hostname=${KUBEFLOW_HOSTNAME}
-  ks param set jupyter jupyterHubAuthenticator iap
-  ks param set pipeline mysqlPd "${DEPLOYMENT_NAME}-storage-pipeline-db"
-  ks param set pipeline nfsPd "${DEPLOYMENT_NAME}-storage-pipeline-nfs"
+  ks param set jupyter jupyterHubAuthenticator iap --env=default
```

Comment:
Why are you setting these parameters only in the environment default?

Reply:
the immediate reason is that i want to be able to persist the configuration if I upgrade the jupyter hub. otherwise it might be lost. 

does it make sense for all customizations(non default parameters) to set to a specific environment?

Reply:
this step would remove the customized parameter
https://github.com/kubeflow/kubeflow/pull/2479/files#diff-9e6f5278480335e834537d84293b656bR32

Reply:
Ok so you're depending on environment specific overrides to keep track of user defined overrides of parameters so you can preserve them on update.

Are you planning on adding an E2E test to verify that behavior to ensure refactorings don't break it?

Reply:
I think another option that might be better would be to define the default parameters inside your .libsonnet file e.g. in your libsonnet

```
params = defaultParams + overrides
```

Don't define optional parameters inside your prototype so parameters will only be defined when the user explicitly decides to override a parameter.

The advantage of this approach is that it doesn't depend on using specific environments so I think its less brittle.

Reply:
I am not sure I entirely follow the idea here, but here are some questions
> define the default parameters inside your .libsonnet file

If so would user not be able to `ks param list` to get the parameter default values?

> `params = defaultParams + overrides`

Where should I store the overrides and ensure the override value can be persist when I delete the component?

Given that we rely on default environment and we are on the way of deprecating ksonnet, it's ok to use default environment as the current approach?

Reply:
> If so would user not be able to `ks param list` to get the parameter default values?
Yes.

> Where should I store the overrides and ensure the override value can be persist when I delete the component?

You would define the default values in the .libsonnet file for pipelines; e.g like we do for tf serving
https://github.com/kubeflow/kubeflow/blob/master/kubeflow/tf-serving/tf-serving.libsonnet#L5

User defined overrides would still be defined in the params.libsonnet files but they wouldn't necessarily be tied to a specific environment. 

> Given that we rely on default environment and we are on the way of deprecating ksonnet, is it ok to use default environment as the current approach?

Up to you but I think the current approach is more brittle. Specifically, you are relying on logic in kfctl to correctly set the parameters for just the default environment so that defaults are still preserved in the components/params.libsonnet file. 

At a minimum you'll also need to reimplement this logic in the kfctl go binary.

The approach I suggest doesn't have that brittleness; and it doesn't rely on users only overriding parameters in a specific environment.

Reply:
thanks for the sample. it's helped me understand the idea. done.